File: Insert-Node/InsertNode.java
class Node {
    int data;
    Node next;

    Node(int val) {
        data = val;
        next = null;
    }
}

public class InsertNode {

    // Function to insert a node at a specific position
    public static Node insertNode(Node head, int data, int position) {
        Node newNode = new Node(data);

        if (position == 1) {  // Insert at the head
            newNode.next = head;
            return newNode;
        }

        Node current = head;
        for (int i = 1; i < position - 1; i++) {
            if (current == null) {
                System.out.println("Invalid position!");
                return head;
            }
            current = current.next;
        }

        if (current == null) {  // Position is beyond the list length
            System.out.println("Invalid position!");
            return head;
        }

        newNode.next = current.next;
        current.next = newNode;

        return head;
    }

    // Function to print the linked list
    public static void printList(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        // Create a linked list: 1 -> 2 -> 3 -> 4 -> 5
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        head.next.next.next.next = new Node(5);

        // Print the original list
        System.out.print("Original list: ");
        printList(head);

        // Insert a new node with value 2.5 at position 3
        head = insertNode(head, 200, 3);

        // Print the modified list
        System.out.print("List after inserting 200 at position 3: ");
        printList(head);

        // Clean up remaining nodes (optional in Java due to garbage collection)
        while (head != null) {
            Node temp = head;
            head = head.next;
            // No explicit delete needed; Java handles memory management
        }
    }
}

File: Insert-Node/insertNode.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


def insert_node(head, data, position):
    new_node = Node(data)

    if position == 1:  # Insert at the head
        new_node.next = head
        return new_node

    current = head
    for _ in range(1, position - 1):
        if current is None:
            print("Invalid position!")
            return head
        current = current.next

    if current is None:  # Position is beyond the list length
        print("Invalid position!")
        return head

    new_node.next = current.next
    current.next = new_node

    return head


def print_list(head):
    current = head
    while current is not None:
        print(current.data, end=" -> ")
        current = current.next
    print("None")


if __name__ == "__main__":
    # Create a linked list: 1 -> 2 -> 3 -> 4 -> 5
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    head.next.next.next = Node(4)
    head.next.next.next.next = Node(5)

    # Print the original list
    print("Original list: ", end="")
    print_list(head)

    # Insert a new node with value 2.5 at position 3
    head = insert_node(head, 2.5, 3)

    # Print the modified list
    print("List after inserting 2.5 at position 3: ", end="")
    print_list(head)

    # Clean up remaining nodes
    while head is not None:
        temp = head
        head = head.next
        del temp
